### DISCUSSION TITLE
How to get result in SignalR filter?

### DISCUSSION THREAD

**Discussion by sasgosh** [dotnet/aspnetcore#52750] [Q&A]
Hello,

I have **SignalR filter** and want to get result after method invocation, for example:

```
public async ValueTask<object> InvokeMethodAsync(HubInvocationContext context, Func<HubInvocationContext, ValueTask<object>> next) 
{
var result = await next(context);
```

But, result is always null (?!). I don't know why.

However, in **minimal api filter**, a can get result with:

```
public async ValueTask<object> InvokeAsync(EndpointFilterInvocationContext context, EndpointFilterDelegate next) 
{
var result = await next(context);

if (result is ProblemHttpResult) 
{
    //do something with it
}

return result;
```

It seems to me that the filters work differently which is quite strange.
Can someone help me how to get the result in the SignalR filter?

**Comment by BrennanConroy** [comment]
Are you returning anything from your hub method?

**Reply by sasgosh** [reply]
Of course. I see the result in Postman. But it is not in the filter.

![Screenshot_1](https://github.com/dotnet/aspnetcore/assets/39879027/99810334-b862-4a1c-97a5-f5b96bc6864d)

**Reply by BrennanConroy** [reply]
A code sample would be helpful to understand what you're doing.

**Reply by sasgosh** [reply]
I want to capture the result of the method in the filter before it is delivered to the user.

Here is demo code:


```
using Microsoft.AspNetCore.SignalR;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddSignalR()
    .AddHubOptions<ChatHub>(options => { options.AddFilter<Filter>(); });

var app = builder.Build();

app.UseHttpsRedirection();
app.MapHub<ChatHub>("");

app.Run();


public class ChatHub : Hub {
    public async Task Test() {
        await Clients.Caller.SendAsync("Test", "Hello message");
    }
}


public class Filter() : IHubFilter {
    public async ValueTask<object> InvokeMethodAsync(HubInvocationContext context, Func<HubInvocationContext, ValueTask<object>> next) {        
        var result = await next(context);

        //here I want to access the result of the method that was called but it is always null
           
        return result;
    }
}
```

**Reply by BrennanConroy** [reply]
You are clearly **not** returning anything from the hub method.

What you want is https://github.com/dotnet/aspnetcore/issues/48401

**Reply by sasgosh** [reply]
I wrote above, **I see the message in Postman**. The client gets the response, the problem is that it cannot be caught through the filter. Apparently SignalR filters only allow management of **input data** (arguments sent by the user) while **output** cannot be captured. Now I don't know if it's a bug or an omission. For example, Api filters work by intercepting both the input before the method, and the result after the method, before delivering it to the user. This guy's proposal is a variation on the theme, he also wants to take the output over before it reaches the user but not as part of the InvokeMethod event. That is also an ok suggestion in my opinion.

**Reply by davidfowl** [reply]
```C#
public async Task Test() 
{
    await Clients.Caller.SendAsync("Test", "Hello message");
}
```

The Test method doesn't return anything in the context of SignalR. That's what @BrennanConroy means when he said your method doesn't return anything.

What are you expecting the data to have?

**Reply by BrennanConroy** [reply]
You are misunderstanding the concept here.
If you take a look at minimal APIs which is what you are comparing to, the equivalent in SignalR is
```csharp
// signalr
public class ChatHub : Hub {
    public string Test() {
        return "Hello message"; // notice we are returning a value
    }
}

// minimal api
app.MapGet("/", () =>
{
    return "Hello message"; // notice we are also returning here
});
```

Now, your SignalR code is not returning. It's writing to a client directly. The equivalent in minimal APIs is
```csharp
app.MapGet("/", (context) =>
{
    // see we aren't returning anything, and are writing directly to the response. This isn't capturable by endpoint filters.
    context.Response.WriteAsync("Hello message");
});
```

**Reply by sasgosh** [reply]
Ah, now things are much clearer. In the official SignalR documentation (picture below), it says that "**await Clients.All.SendAsync**" returns data to the user. And it works, **but it bypasses the Filter**. Now you are saying that I should use "**return**" to return the data so that I can access it through the Filter? So how can I control who the data is returned to, all users, group or caller?

![Screenshot_2](https://github.com/dotnet/aspnetcore/assets/39879027/d2bb587a-e4c6-4597-9a11-80797656094b)

**Reply by davidfowl** [reply]
>  And it works, but it bypasses the Filter. Now you are saying that I should use "return" to return the data so that I can access it through the Filter? So how can I control who the data is returned to, all users, group or caller?

This is why Brennan pointed you at the other issue. There are no filters for outgoing messages from server to client.

**Reply by sasgosh** [reply]
Ok, but what is differance between "await Clients.All.SendAsync" and "return"? In both case user get its message.

**Reply by davidfowl** [reply]
When you call a hub method from the client, that method can have a return value (as pointed out here https://github.com/dotnet/aspnetcore/discussions/52750#discussioncomment-7833358). Calling `await Clients.All.SendAsync` isn't a return value, it's a server to client method invocation. Nothing is being "returned", instead, you are calling the "ReceiveMessage" method on the client side. That method itself has no return value, and it's not even possible in SignalR to get a return value from using All/Group etc.

**Reply by sasgosh** [reply]
Just to see if I got it right

**When I use "return"**
I can catch output in Filter
I cannot send to all or group
I cannot invoke client method

Client get this:
{"type":3,"invocationId":"0","result":"Hello message"}


**When I use "await Clients..."**
I can send message to caller, group and all
I can invoke method on client
But I cannot catch output in Filter?!

Client get this:
{"type":1,"target":"Test","arguments":["Hello message"]}

**Reply by davidfowl** [reply]
Right, we support filters for client to server invocations, but not server to client invocations.

**Comment by sasgosh** [comment]
Thanks to everyone for the replies, things are much clearer now. My suggestion is that the filters should be improved in the future and that the response of client method invocation can be intercepted. Also, the official SignalR documentation should be also improved because it doesn't clearly state that there are two types of responses (with return and client method invocation) and that some can be intercepted by filters while some can't.

**Reply by davidfowl** [reply]
We'll update the docs, it has nothing to do with responses, and everything to do with client method invocation.
